¡Hola!

Mi nombre es Marcos Torres y tengo el gusto de revisar tu código el día de hoy.

Cuando vea algo notable o algún asunto en el notebook, te dejaré un comentario o un hint. Se que encontraras la mejor respuesta para resolver todos los comentarios, de no ser así, no te preocupes en futuras iteraciones dejaré comentarios y pistas más específicos.

Encontrarás comentarios en verde, amarillo o rojo como los siguientes:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Buen trabajo. ¡Lo hiciste muy bien!
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Nota. Se puede mejorar.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesitas corregirlo. Este bloque indica que se requiere una correción. El trabajo no se acepta si tiene estos bloques.
</div>

Puedes responder a mis comentarios usando estos bloques:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante.</b> <a class="tocSkip"></a>
</div>

## En este proyecto encontraremos los mejores lugares donde abrir 200 pozos nuevos de petróleo. Creando un modelo que ayude a elegir la región con el mayor margen de beneficio. Analizando los beneficios y riesgos potenciales utilizando la técnica bootstrapping.

se importan las librerias necesarias

In [32]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

se cargan los datos y se preprocesan

In [33]:
def cargar_datos(archivo):
    df = pd.read_csv(archivo)
    return df

def preprocesar_datos(df):
    df = df.dropna()
    return df

Ahora se entrena el modelo de regresion lineal

In [34]:
def entrenar_modelo(df):
    X = df[['f0', 'f1', 'f2']]
    y = df['product']
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.25, random_state=42)
    
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    y_pred = model.predict(X_val)
    
    rmse = mean_squared_error(y_val, y_pred, squared=False)
    print("RMSE:", rmse)
    return y_val, y_pred

Ahora creamos una funcion para aplicar la tecnica bootstrapping a las 3 regiones 

In [35]:
def analisis_bootstrapping(df_region, revenue_per_unit=4500, cost_per_well=500000, n_iterations=1000):
    bootstrap_results = []
    
    for _ in range(n_iterations):
        sample_500 = df_region.sample(n=500, replace=True) 
        df_top_200 = sample_500.sort_values(by='pred', ascending=False).head(200)
        df_top_200['real_income'] = df_top_200['real'] * revenue_per_unit  
        df_top_200['real_profit'] = df_top_200['real_income'] - cost_per_well  
        bootstrap_results.append(df_top_200['real_profit'].sum())
    
    bootstrap_results = np.array(bootstrap_results)
    mean_profit = bootstrap_results.mean()
    conf_interval = np.percentile(bootstrap_results, [2.5, 97.5])
    loss_risk = np.mean(bootstrap_results < 0)  
    
    print(f"Beneficio promedio (Bootstrap): {mean_profit}")
    print(f"Intervalo de confianza del 95%: {conf_interval}")
    print(f"Riesgo de pérdida: {loss_risk * 100}%")
    
    return mean_profit, conf_interval, loss_risk


definimos 'archivos' que es donde estan los datos para las 3 regiones

In [36]:
archivos = ['/datasets/geo_data_0.csv', '/datasets/geo_data_1.csv', '/datasets/geo_data_2.csv']

Ahora analizamos las 3 regiones de extraccion de petroleo utilizando un bucle for

In [37]:
for i, archivo in enumerate(archivos):
    print(f"Analizando región {i}...")
    df = cargar_datos(archivo)
    df = preprocesar_datos(df)
    real, pred = entrenar_modelo(df)
    df_region = pd.DataFrame({'real': real, 'pred': pred})
    analisis_bootstrapping(df_region)
    print("\n------------------------\n")

Analizando región 0...
RMSE: 37.756600350261685
Beneficio promedio (Bootstrap): 3918108.2938534175
Intervalo de confianza del 95%: [-1190655.84592237  9623790.69074698]
Riesgo de pérdida: 6.5%

------------------------

Analizando región 1...
RMSE: 0.890280100102884
Beneficio promedio (Bootstrap): 4476183.3459137175
Intervalo de confianza del 95%: [ 379692.38270072 8401604.86401842]
Riesgo de pérdida: 1.7999999999999998%

------------------------

Analizando región 2...
RMSE: 40.14587231134218
Beneficio promedio (Bootstrap): 3912541.468444423
Intervalo de confianza del 95%: [-910312.29710672 8998887.63520711]
Riesgo de pérdida: 6.9%

------------------------



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Buen trabajo. Gracias por atender a los comentarios, ahora el bootstrap se ha realizado de manera correcta.
</div>

## Ya que las 3 regiones han sido analizadas por el modelo, podemos observar que la region 0 y la region 2 dan un % de riesgo de perdida del 6.5 y 6.9 respectivamente mientras que en la region 1 hay un riesgo de perdida de solo el 1.7999%, ademas el RMSE en la region 1 es de .89 lo cual indica que es mas preciso y confiable en comparacion con las regiones 0 y 2, y por ultimo podemos observar que de las 3 regiones el beneficio promedio es mayor en la region 1.
## Entonces la region 1 es la mejor region de las 3 para abrir 200 pozos nuevos de petroleo.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Buenas conclusiones. Basadas en los resultados.
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

¡Hola!
    
Te quiero felicitar por tu proyecto lograste realizar la lógica del bootstrap para encontrar la probabilidad de riesgo de pérdida en cada región, te deseo mucho éxito en los siguientes sprints.
    
Saludos, Marcos.
</div>